Make substring null behavior congruent with tsql

jusjosj opened the pull request:
### Description

This commit addresses a difference in the way that substring handled null arguments between babelfish and tsql.  Prior to this change, babelfish would raise an error on any null input to the substring function regardless of whether or not this is a null constant or another data type with the value of null.  After this change, babelfish will only raise an error on values which are considered to be null constants restoring paraity to tsql
 
Task:BABEL-3347
Signed-off-by: Justin Jossick [jusjosj@amazon.com](mailto:jusjosj@amazon.com)

### Issues Resolved

[BABEL-3347] 

### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

2jungkook commented on contrib/babelfishpg_tsql/src/string.c:
```diff
@@ -478,17 +478,32 @@ prepare_format_string(StringInfo buf, char *msg_string, int nargs,
-		ereport(ERROR,
-				(errcode(ERRCODE_SUBSTRING_ERROR),
-					errmsg("Argument data type NULL is invalid for argument 1 of substring function")));
+		fe = (FuncExpr *)fcinfo->flinfo->fn_expr;
```
To be safer, let's add some null/NIL checks, e.g., `fe` and `fe->args`.

2jungkook commented on contrib/babelfishpg_tsql/src/string.c:
```diff
@@ -478,17 +478,32 @@ prepare_format_string(StringInfo buf, char *msg_string, int nargs,
-				(errcode(ERRCODE_SUBSTRING_ERROR),
-					errmsg("Argument data type NULL is invalid for argument 1 of substring function")));
+		fe = (FuncExpr *)fcinfo->flinfo->fn_expr;
+		node = (Node *)fe->args->elements[0].ptr_value;
```
Can we use a wrapper function in pg_list.h, e.g., `linitial_node`?

2jungkook commented on contrib/babelfishpg_tsql/src/string.c:
```diff
@@ -478,17 +478,32 @@ prepare_format_string(StringInfo buf, char *msg_string, int nargs,
+						errmsg("Argument data type NULL is invalid for argument 1 of substring function")));
+		}
+		/* Handle other cases for null such as table variables */
+		PG_RETURN_NULL();
```
Could you also add a test for `cast(null as varchar)` case?

2jungkook commented on contrib/babelfishpg_tsql/src/string.c:
```diff
@@ -478,17 +478,32 @@ prepare_format_string(StringInfo buf, char *msg_string, int nargs,
 	}
+
 	if (PG_ARGISNULL(1) || PG_ARGISNULL(2))
 		PG_RETURN_NULL();
```
Please add a test where the 2nd or the 3rd argument is null from table variables.

jusjosj replied:
I can, but this is already covered in the tests for babel-1114 so I'm not sure that it is necessary

2jungkook commented on test/JDBC/expected/BABEL-3347-vu-prepare.out:
```diff
@@ -0,0 +1,11 @@
+CREATE TABLE babel_3347_table(text_test VARCHAR(100));
```
If possible, let's add this test to other version's schedule files.

2jungkook commented on test/JDBC/expected/BABEL-3347-vu-verify.out:
```diff
@@ -0,0 +1,26 @@
+-- table variable
+SELECT SUBSTRING(text_test, 2,3) FROM babel_3347_table
```
The one of the main purposes of version upgrade test is to make sure all objects are correctly migrated. Please create a view for each `SELECT` statement.

2jungkook commented on contrib/babelfishpg_tsql/src/string.c:
```diff
@@ -478,17 +478,40 @@ prepare_format_string(StringInfo buf, char *msg_string, int nargs,
-					errmsg("Argument data type NULL is invalid for argument 1 of substring function")));
+		fe = (FuncExpr *)fcinfo->flinfo->fn_expr;
+
+		if (fe == NULL || fe->args == NULL) {
```
For a list, we use `NIL` instead of `NULL`.

2jungkook commented on contrib/babelfishpg_tsql/src/string.c:
```diff
@@ -478,17 +478,40 @@ prepare_format_string(StringInfo buf, char *msg_string, int nargs,
+
+		if (fe == NULL || fe->args == NULL) {
+			ereport(ERROR,
+					(errcode(ERRCODE_SUBSTRING_ERROR),
```
Let's use `ERRCODE_INTERNAL_ERROR`.

2jungkook commented on test/JDBC/expected/BABEL-3347-vu-cleanup.out:
```diff
@@ -0,0 +1,26 @@
+DROP VIEW IF EXISTS babel_3347_vu1
```
Here, `vu` means `version_upgrade`. So the naming convention would be `babel_3347_vu_v1`.

jusjosj committed:
[BABEl-3347] Allow nulls as first argument to substring

This changes the way in which null is handled by the substring function
for the first input.  Initially we raised an error, however this was a
major difference compared to TSQL which only raises an error on the null
literal.  With this commit, we will return null instead of raising an
error on null inputs

jusjosj committed:
[BABEL-3347] fix tests

jusjosj committed:
[BABEL-3347] Use parser to detect null first args in substring function

Since tsql appears to reject nulls at the parser level for substring
babelfish should as well. This change allows us to abort batch
statements where substring is called with a null literal the same was as
is done in tsql

jusjosj committed:
[BABEL-3347] Correct before tests

jusjosj committed:
[BABEL-3347] more test fixes

jusjosj committed:
[BABEL-3347] more test clean up for upgrade tests

jusjosj committed:
[BABEL-3347] Refactor parser check

jusjosj committed:
[BABEL-3347] use correct datatype in test

2jungkook commented on contrib/babelfishpg_tsql/src/tsqlIface.cpp:
```diff
@@ -2153,6 +2153,29 @@ class tsqlMutator : public TSqlParserBaseListener
+			{
+				auto id = fpnsds->id().back();
+
+				if (id->keyword()->SUBSTRING()) /* NULLIF */
```
1. Please fix the comment
2. We need to fix NULLIF (and probably for CHECKSUM) as well. Please create a JIRA for that.

jusjosj committed:
[BABEL-3347] Fix comment
